File: 2025-01-13/Merge_Two_Sorted_Lists.java
// 21. Merge Two Sorted Lists

// https://leetcode.com/problems/merge-two-sorted-lists/description

public class Merge_Two_Sorted_Lists {
    public static void printList(ListNode head) {
        ListNode current = head;
        while (current != null) {
            System.out.print(current.val + (current.next != null ? " -> " : ""));
            current = current.next;
        }
    }
    public static ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode temp = new ListNode();
        ListNode ans = temp;

        while(list1 != null && list2 != null) {
            if(list1.val <= list2.val) {
                ans.next = list1;
                list1 = list1.next;
            }
            else {
                ans.next = list2;
                list2 = list2.next;
            }

            ans = ans.next;
        }

        ans.next = (list1 != null) ? list1 : list2;

        return temp.next;
    }
    public static void main(String[] args) {
        ListNode list1 = new ListNode(1, new ListNode(2, new ListNode(4)));

        // Create list2: [1,3,4]
        ListNode list2 = new ListNode(1, new ListNode(3, new ListNode(4)));

        // Print the list to verify
        printList(list1);
        printList(list2);
        ListNode mergedList = mergeTwoLists(list1, list2);
        System.out.println();
        printList(mergedList);

    }
}

File: 2026-02-01/Merge_Two_Sorted_Lists.py
from typing import Optional
from Linked_List import Linked_List, ListNode


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(0)
        cur = head

        while list1 and list2:
            if list1.val <= list2.val:
                cur.next = list1
                list1 = list1.next
            else:
                cur.next = list2
                list2 = list2.next

            cur = cur.next

        cur.next = list1 or list2

        return head.next


sol = Solution()
ll = Linked_List()
head1 = ll.build_from_list([1,2,4])
head2 = ll.build_from_list([1,3,4])
ll.print_list(sol.mergeTwoLists(head1, head2))